controller_manager:
  ros__parameters:

    update_rate: 800  # Match your /clock rate
    use_sim_time: true

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    io_and_status_controller:
      type: ur_controllers/GPIOController

    speed_scaling_state_broadcaster:
      type: ur_controllers/SpeedScalingStateBroadcaster

    force_torque_sensor_broadcaster:
      type: ur_controllers/ForceTorqueStateBroadcaster

    joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

    scaled_joint_trajectory_controller:
      type: ur_controllers/ScaledJointTrajectoryController

    forward_velocity_controller:
      type: velocity_controllers/JointGroupVelocityController

    forward_position_controller:
      type: position_controllers/JointGroupPositionController

    gripper_position_controller:
      type: position_controllers/GripperActionController


joint_state_broadcaster:
  ros__parameters:
    use_sim_time: true
    publish_rate: 80.0
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
      - finger_joint

force_torque_sensor_broadcaster:
  ros__parameters:
    sensor_name: tcp_fts_sensor
    state_interface_names:
      - force.x
      - force.y
      - force.z
      - torque.x
      - torque.y
      - torque.z
    frame_id: tool0
    topic_name: ft_data


joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 100.0
    action_monitor_rate: 20.0
    # allow_partial_joints_goal: false
    constraints:
      stopped_velocity_tolerance: 0.5
      goal_time: 5.0
      shoulder_pan_joint: { trajectory: 0.8, goal: 0.35 }
      shoulder_lift_joint: { trajectory: 0.8, goal: 0.35 }
      elbow_joint: { trajectory: 0.8, goal: 0.35 }
      wrist_1_joint: { trajectory: 0.8, goal: 0.45 }
      wrist_2_joint: { trajectory: 0.8, goal: 0.35 }
      wrist_3_joint: { trajectory: 0.8, goal: 0.35 }


scaled_joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 100.0
    action_monitor_rate: 20.0
    allow_partial_joints_goal: false
    constraints:
      stopped_velocity_tolerance: 0.3
      goal_time: 2.0
      shoulder_pan_joint: { trajectory: 0.25, goal: 0.20 }
      shoulder_lift_joint: { trajectory: 0.25, goal: 0.20 }
      elbow_joint: { trajectory: 0.25, goal: 0.20 }
      wrist_1_joint: { trajectory: 0.30, goal: 0.25 }
      wrist_2_joint: { trajectory: 0.25, goal: 0.20 }
      wrist_3_joint: { trajectory: 0.25, goal: 0.20 }

forward_velocity_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    interface_name: velocity

forward_position_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint


gripper_position_controller:
  ros__parameters:
    action_monitor_rate: 20.0
    allow_stalling: true
    goal_tolerance: 0.01
    joint: 'finger_joint'
    max_effort: 100.0
    stall_timeout: 3.0
    stall_velocity_threshold: 0.001
